File: 2020/aoc-elixir/lib/advent_of_code/day_04.ex
defmodule AdventOfCode.Day04 do
  def input_to_maps(input) do
    String.split(input, "\n\n")
    |> Enum.map(&(Regex.scan(~r/([[:alpha:]]+):([[:alnum:]#]+)/, &1)
                  |> Map.new(fn [_, k, v] -> {String.to_atom(k), v} end)))
  end

  def has_keys(map) do
    [:byr, :iyr, :eyr, :hgt, :hcl, :ecl, :pid]
    |> Enum.all?(&(Map.has_key?(map, &1)))
  end

  def part1(args) do
    Enum.filter(input_to_maps(args), &has_keys/1) |> Enum.count
  end

  def val_byr(x), do: (1920 <= String.to_integer(x)) and (String.to_integer(x) <= 2002)
  def val_iyr(x), do: (2010 <= String.to_integer(x)) and (String.to_integer(x) <= 2020)
  def val_eyr(x), do: (2020 <= String.to_integer(x)) and (String.to_integer(x) <= 2030)
  def val_hgt([_, x, "cm"]), do: (150 <= String.to_integer(x)) and (String.to_integer(x) <= 193)
  def val_hgt([_, x, "in"]), do: (59 <= String.to_integer(x)) and (String.to_integer(x) <= 76)
  def val_hgt(_), do: false
  def val_hcl(x), do: Regex.match?(~r/#[0-9a-f]{6}/, x)
  def val_ecl(x), do: Enum.member?(["amb", "blu", "brn", "gry", "grn", "hzl", "oth"], x)
  def val_pid(x), do: Regex.match?(~r/^\d{9}$/, x)

  def val_passport(%{byr: byr, iyr: iyr, eyr: eyr, hgt: hgt, hcl: hcl, ecl: ecl, pid: pid}) do
    Enum.all?([val_byr(byr), val_iyr(iyr), val_eyr(eyr), val_hgt(Regex.run(~r/(\d+)(cm|in)/, hgt)),
               val_hcl(hcl), val_ecl(ecl), val_pid(pid)])
  end
  def val_passport(_), do: false

  def part2(args) do
    input_to_maps(args) |> Enum.filter(&val_passport/1) |> Enum.count
  end
end

File: 2020/aoc-clojure-2020/src/aoc_clojure_2020/day-04.clj
(ns aoc-clojure-2020.day-04
  (:require [aoc-clojure-2020.helpers :refer [get-input]]
            [clojure.string :as s]
            [clojure.spec.alpha :as v]))

(def input (as-> (get-input 4) i
             (s/split i #"\n\n")
             (map #(->> (s/split % #"[\ \n]+")
                        (map (fn [x] (let [[k v] (s/split x #":")]
                                       [(keyword (str *ns*) k) v])))
                        (into {})) i)))

(v/def ::byr #(<= 1920 (Integer/parseInt %) 2002))
(v/def ::iyr #(<= 2010 (Integer/parseInt %) 2020))
(v/def ::eyr #(<= 2020 (Integer/parseInt %) 2030))
(v/def ::hgt (v/or :cm #(if-let [[_ h] (re-matches #"(\d+)cm" %)] (<= 150 (Integer/parseInt h) 193))
                   :in #(if-let [[_ h] (re-matches #"(\d+)in" %)] (<= 59 (Integer/parseInt h) 76))))
(v/def ::hcl #(re-matches #"#[0-9a-f]{6}" %))
(v/def ::ecl #{"amb" "blu" "brn" "gry" "grn" "hzl" "oth"})
(v/def ::pid #(re-matches #"\d{9}" %))

(v/def ::passport (v/keys :req [::byr ::iyr ::eyr ::hgt ::hcl ::ecl ::pid]
                          :opt [::cid]))

(def part-1 (-> (filter #(every? % [::byr ::iyr ::eyr ::hgt ::hcl ::ecl ::pid]) input)
                count))

(def part-2 (-> (filter #(v/valid? ::passport %) input)
                count))
part-2
